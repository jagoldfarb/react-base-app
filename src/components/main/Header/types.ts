import { IUser } from 'types/user.types';

export interface IHeaderProps {
  isMenuOpened: boolean;
  onLogoutClick: () => void;
  toggleMenu: () => void;
  user: null | IUser;
}
