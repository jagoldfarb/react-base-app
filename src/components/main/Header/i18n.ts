import i18next from 'i18next';
import { locales } from 'configs/i18n';

i18next.addResources(locales.EN, 'HEADER', {
  LOGIN: 'Log In',
  LOGOUT: 'Log Out',
  SIGNUP: 'Sign Up',
});
