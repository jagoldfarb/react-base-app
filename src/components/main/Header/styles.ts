import styled, { css } from 'styled-components';

import { HEADER_HEIGHT } from 'helpers/dimensions';
import { COLORS } from 'helpers/enums/colors';
import { SIZES } from 'helpers/enums/sizes';

import { ReactComponent as OpenMenuSVG } from './assets/open-menu.svg';
import { ReactComponent as CloseMenuSVG } from './assets/close-menu.svg';

export const HeaderContainer = styled.div`
  position: fixed;
  width: 100%;
  height: ${HEADER_HEIGHT}px;
  background-color: ${COLORS.white};
  z-index: 4;
  box-shadow: 0px 3px 6px ${COLORS.boxShadow};
`;

export const HeaderContent = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  height: 100%;
  max-width: ${SIZES.smallDesktop}px;
  margin: 0 auto;

  @media (max-width: ${SIZES.smallDesktop}px) {
    padding: 0 25px;
  }
`;

export const MenuOverlay = styled.div`
  @media (max-width: ${SIZES.mediumTablet}px) {
    ${({ showOverlay }: { showOverlay: boolean }) =>
      showOverlay &&
      css`
        position: fixed;
        width: 100%;
        height: 100%;
        top: 0;
        left: 0;
        background: ${COLORS.boxShadow};
      `}
  }
`;

export const MenuContainer = styled.div`
  display: contents;
`;

export const MenuIconContainer = styled.div`
  display: none;

  @media (max-width: ${SIZES.mediumTablet}px) {
    display: flex;
    z-index: 3;
  }
`;

export const MenuIcon = css`
  cursor: pointer;
`;

export const OpenMenuIcon = styled(OpenMenuSVG)`
  ${MenuIcon}
`;

export const CloseMenuIcon = styled(CloseMenuSVG)`
  ${MenuIcon}
`;

export const NavLinksContainer = styled.ul`
  display: flex;

  li {
    display: inline-block;
    padding: 0 25px;

    &:last-child {
      padding-right: 0;
    }
  }

  li a,
  li span {
    color: ${COLORS.black};
    font-size: 14px;
    cursor: pointer;
  }

  li a.active {
    color: ${COLORS.greenChristi};
  }

  @media (max-width: ${SIZES.mediumTablet}px) {
    position: fixed;
    top: 0;
    right: ${({ showSideMenu }: { showSideMenu: boolean }) => (showSideMenu ? '0' : '-300px')};
    flex-direction: column;
    width: 300px;
    height: 100vh;
    margin: 0;
    padding: 70px 0px;
    background-color: ${COLORS.white};
    transition: all 0.3s ease-in-out;

    li {
      padding: 15px 0;
      text-align: center;
    }
  }
`;
