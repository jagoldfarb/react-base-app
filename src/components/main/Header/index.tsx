import React, { FunctionComponent } from 'react';
import { NavLink } from 'react-router-dom';
import i18next from 'i18next';

import Logo from 'components/commons/Logo';

import { IHeaderProps } from './types';
import {
  HeaderContainer,
  HeaderContent,
  MenuOverlay,
  MenuContainer,
  MenuIconContainer,
  OpenMenuIcon,
  CloseMenuIcon,
  NavLinksContainer,
} from './styles';
import './i18n';

const Header: FunctionComponent<IHeaderProps> = (props: IHeaderProps) => {
  const { isMenuOpened, onLogoutClick, toggleMenu, user } = props;

  const MenuIcon = isMenuOpened ? CloseMenuIcon : OpenMenuIcon;

  return (
    <HeaderContainer data-testid="header">
      <HeaderContent>
        <Logo />
        <MenuOverlay data-testid="menu-overlay" onClick={toggleMenu} showOverlay={isMenuOpened} />
        <MenuContainer data-testid="menu-content">
          <MenuIconContainer>
            <MenuIcon onClick={toggleMenu} />
          </MenuIconContainer>
          <NavLinksContainer showSideMenu={isMenuOpened}>
            <li>
              <NavLink
                activeClassName="active"
                data-testid="nav-link-1"
                onClick={toggleMenu}
                to="/link-1"
              >
                Link 1
              </NavLink>
            </li>
            <li>
              <NavLink
                activeClassName="active"
                data-testid="nav-link-2"
                onClick={toggleMenu}
                to="/link-2"
              >
                Link 2
              </NavLink>
            </li>
            <li>
              <NavLink
                activeClassName="active"
                data-testid="nav-link-3"
                onClick={toggleMenu}
                to="/link-3"
              >
                Link 3
              </NavLink>
            </li>
            {!user ? (
              <>
                <li>
                  <NavLink
                    activeClassName="active"
                    data-testid="nav-link-signup"
                    onClick={toggleMenu}
                    to="/signup"
                  >
                    {i18next.t('HEADER:SIGNUP')}
                  </NavLink>
                </li>
                <li>
                  <NavLink
                    activeClassName="active"
                    data-testid="nav-link-login"
                    onClick={toggleMenu}
                    to="/login"
                  >
                    {i18next.t('HEADER:LOGIN')}
                  </NavLink>
                </li>
              </>
            ) : (
              <li>
                <span data-testid="nav-link-logout" onClick={onLogoutClick} role="presentation">
                  {i18next.t('HEADER:LOGOUT')}
                </span>
              </li>
            )}
          </NavLinksContainer>
        </MenuContainer>
      </HeaderContent>
    </HeaderContainer>
  );
};

export default Header;
