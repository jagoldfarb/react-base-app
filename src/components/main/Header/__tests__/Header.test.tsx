import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { userFactory } from 'tests/user.factory';

import Header from '..';

const defaultProps = {
  isMenuOpened: false,
  toggleMenu: jest.fn(),
  user: null,
};

describe('Header component', () => {
  it('Should render the header', () => {
    const { asFragment, getByTestId, queryByTestId } = render(
      <MemoryRouter>
        <Header {...defaultProps} />
      </MemoryRouter>,
    );
    expect(getByTestId('header')).toBeDefined();
    expect(getByTestId('menu-content')).toBeDefined();
    expect(getByTestId('nav-link-1')).toBeDefined();
    expect(getByTestId('nav-link-2')).toBeDefined();
    expect(getByTestId('nav-link-3')).toBeDefined();
    expect(getByTestId('nav-link-signup')).toBeDefined();
    expect(getByTestId('nav-link-login')).toBeDefined();
    expect(queryByTestId('nav-link-logout')).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the header with menu opened', () => {
    const props = { ...defaultProps, isMenuOpened: true };
    const { asFragment, getByTestId, queryByTestId } = render(
      <MemoryRouter>
        <Header {...props} />
      </MemoryRouter>,
    );
    expect(getByTestId('header')).toBeDefined();
    expect(getByTestId('menu-overlay')).toBeDefined();
    expect(getByTestId('menu-content')).toBeDefined();
    expect(getByTestId('nav-link-1')).toBeDefined();
    expect(getByTestId('nav-link-2')).toBeDefined();
    expect(getByTestId('nav-link-3')).toBeDefined();
    expect(getByTestId('nav-link-signup')).toBeDefined();
    expect(getByTestId('nav-link-login')).toBeDefined();
    expect(queryByTestId('nav-link-logout')).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the header with user logged in', () => {
    const props = { ...defaultProps, user: userFactory(1) };
    const { asFragment, getByTestId, queryByTestId } = render(
      <MemoryRouter>
        <Header {...props} />
      </MemoryRouter>,
    );
    expect(getByTestId('header')).toBeDefined();
    expect(getByTestId('menu-overlay')).toBeDefined();
    expect(getByTestId('menu-content')).toBeDefined();
    expect(getByTestId('nav-link-1')).toBeDefined();
    expect(getByTestId('nav-link-2')).toBeDefined();
    expect(getByTestId('nav-link-3')).toBeDefined();
    expect(queryByTestId('nav-link-signup')).toBeNull();
    expect(queryByTestId('nav-link-login')).toBeNull();
    expect(getByTestId('nav-link-logout')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });
});
