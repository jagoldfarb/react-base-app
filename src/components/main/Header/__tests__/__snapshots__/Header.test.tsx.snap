// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Should render the header 1`] = `
<DocumentFragment>
  <div
    class="sc-gsTCUz isGlRi"
    data-testid="header"
  >
    <div
      class="sc-dlfnbm sofHW"
    >
      <div
        class="sc-bdfBwQ fajWDP"
        data-testid="logo"
      >
        <a
          href="/home"
        >
          Logo
        </a>
      </div>
      <div
        class="sc-hKgILt MKpTd"
        data-testid="menu-overlay"
      />
      <div
        class="sc-eCssSg laWbwL"
        data-testid="menu-content"
      >
        <div
          class="sc-jSgupP dgpcrU"
        >
          <svg
            class="sc-gKsewC gPMeTF"
          >
            open-menu.svg
          </svg>
        </div>
        <ul
          class="sc-fubCfw cmWXKT"
        >
          <li>
            <a
              data-testid="nav-link-1"
              href="/link-1"
            >
              Link 1
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link-2"
              href="/link-2"
            >
              Link 2
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link-3"
              href="/link-3"
            >
              Link 3
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link-signup"
              href="/signup"
            >
              Sign Up
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link-login"
              href="/login"
            >
              Log In
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Header component Should render the header with menu opened 1`] = `
<DocumentFragment>
  <div
    class="sc-gsTCUz isGlRi"
    data-testid="header"
  >
    <div
      class="sc-dlfnbm sofHW"
    >
      <div
        class="sc-bdfBwQ fajWDP"
        data-testid="logo"
      >
        <a
          href="/home"
        >
          Logo
        </a>
      </div>
      <div
        class="sc-hKgILt geDgRy"
        data-testid="menu-overlay"
      />
      <div
        class="sc-eCssSg laWbwL"
        data-testid="menu-content"
      >
        <div
          class="sc-jSgupP dgpcrU"
        >
          <svg
            class="sc-iBPRYJ jkRnWh"
          >
            close-menu.svg
          </svg>
        </div>
        <ul
          class="sc-fubCfw dBpaIJ"
        >
          <li>
            <a
              data-testid="nav-link-1"
              href="/link-1"
            >
              Link 1
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link-2"
              href="/link-2"
            >
              Link 2
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link-3"
              href="/link-3"
            >
              Link 3
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link-signup"
              href="/signup"
            >
              Sign Up
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link-login"
              href="/login"
            >
              Log In
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Header component Should render the header with user logged in 1`] = `
<DocumentFragment>
  <div
    class="sc-gsTCUz isGlRi"
    data-testid="header"
  >
    <div
      class="sc-dlfnbm sofHW"
    >
      <div
        class="sc-bdfBwQ fajWDP"
        data-testid="logo"
      >
        <a
          href="/home"
        >
          Logo
        </a>
      </div>
      <div
        class="sc-hKgILt MKpTd"
        data-testid="menu-overlay"
      />
      <div
        class="sc-eCssSg laWbwL"
        data-testid="menu-content"
      >
        <div
          class="sc-jSgupP dgpcrU"
        >
          <svg
            class="sc-gKsewC gPMeTF"
          >
            open-menu.svg
          </svg>
        </div>
        <ul
          class="sc-fubCfw cmWXKT"
        >
          <li>
            <a
              data-testid="nav-link-1"
              href="/link-1"
            >
              Link 1
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link-2"
              href="/link-2"
            >
              Link 2
            </a>
          </li>
          <li>
            <a
              data-testid="nav-link-3"
              href="/link-3"
            >
              Link 3
            </a>
          </li>
          <li>
            <span
              data-testid="nav-link-logout"
              role="presentation"
            >
              Log Out
            </span>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
